### ISSUE TITLE
HPACK Decoder's maxHeadesLength

### ISSUE THREAD

**Issue by buchgr** [netty/netty#5774]
A lot of unrelated h2spec tests fail due to exceeding [HPACK Decoder's maxHeadersLength](https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/Decoder.java#L76). It's set to 8KiB by [default](https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2CodecUtil.java#L106).

It's important to note that this is not the maximum size of the dynamic table, that's set via `SETTINGS_HEADER_TABLE_SIZE`. 

From looking through the HPACK code, it seems that `maxHeadersLength` is simple the maximum number of bytes of a decoded header block (from static table + dynamic table) before decoding fails.  Even if the `SETTINGS_HEADER_TABLE_SIZE` is negotiated/set larger than the static `maxHeadersLength`, decoding will fail. As `maxHeadersLength` seems to be the absolute upper bound.

The `maxHeadersLength` was probably introduced to limit the per-stream memory. However, if an endpoint opts-in for a large dynamic table size I think it should be able to do so. So it seems to me, that `maxHeadersLength` should really be `maxHeaderTableSize` + `some max. overhead for static table`. With `maxHeaderTableSize` being the value set via `SETTINGS_HEADER_TABLE_SIZE` and `some max. overhead for static table` being a static value, that can be set in the decoder's constructor.

Thoughts?

@Scottmitch @nmittler @normanmaurer @ejona86

**Comment by nmittler** [comment]
@buchgr sgtm.  Want to throw together a PR to see what it looks like?

**Comment by buchgr** [comment]
Actually, I just realized that my understanding was wrong.

In the HTTP/2 spec I found `SETTINGS_MAX_HEADER_LIST_SIZE`, that is supposed to tell an endpoint the maximum size of a headers block the other endpoint is ready to receive. The default value is unlimited. However, it's only an advisory setting. So it's legal for an endpoint to ignore it. However, it's also legal for a receiving endpoint to ignore it's own advise and enforce a lower limit than what it advised. Which is what we are doing right now - advise unlimited header block size, but enforce a 8KiB header block size. So it seems like our code is correct and legal after all.

We may choose to be a good endpoint and include `SETTINGS_MAX_HEADER_LIST_SIZE` in the initial settings frame by default ... but not sure if that's valuable.
